src/expression.rs

```diff
@@ -75,91 +78,123 @@ fn calculate_function(
     }
 }
 
-pub type EvaluationEnvironment = HashMap<String, num_complex::Complex64>;
-
 impl Expression {
     /// Consume the expression, simplifying it as much as possible using the values provided in the environment.
-    /// If variables are used in the expression which are not present in the environment, evaluation stops there,
-    /// returning the possibly-simplified expression.
-    pub fn evaluate(self, environment: &EvaluationEnvironment) -> Self {
+    pub fn simplify(self) -> Self {
         use Expression::*;
-        match self {
+
+        let simplified = match self {
             FunctionCall {
                 function,
                 expression,
-            } => {
-                let evaluated = (*expression).evaluate(environment);
-                match &evaluated {
-                    Number(value) => Number(calculate_function(&function, value)),
-                    PiConstant => Number(calculate_function(&function, &real!(PI))),
-                    _ => FunctionCall {
-                        function,
-                        expression: Box::new(evaluated),
-                    },
-                }
-            }
+            } => FunctionCall {
+                function,
+                expression: Box::new(expression.simplify()),
```

Comment:
Maybe include the `calculate_function(&function, expression.simplify())` in here?

Reply:
As in: if the expression is/simplifies to a number or π